lambdajs/src/SharedAbstract.hs

```diff
@@ -0,0 +1,262 @@
+
+typeEvalUnOp_ :: (ArrowFail String c, ArrowChoice c) => c (Op, Type) Type
+typeEvalUnOp_ = proc (op, vals) -> case (op, vals) of
+    -- number operator
+    (OToInteger, TNumber) -> returnA -< TNumber
+    (OToInt32, TNumber)   -> returnA -< TNumber
+    (OToUInt32, TNumber)  -> returnA -< TNumber
+    -- boolean operator
+    (OBNot, TBool)        -> returnA -< TBool
+    -- string operators
+    (OStrLen, TString)    -> returnA -< TNumber
+    -- isPrimitive operator
+    (OIsPrim, _)          -> returnA -< TBool
+    -- primToNum operator
+    -- #todo object conversions -> valueOf call
+    (OPrimToNum, _)       -> returnA -< TNumber
+    -- primToStr operator
+    (OPrimToStr, _)       -> returnA -< TString
+    -- primToBool operator
+    (OPrimToBool, _)      -> returnA -< TBool
+    -- typeOf operator
+    (OTypeof, _)          -> returnA -< TString
+    -- math operators
+    (OMathExp, TNumber)   -> returnA -< TNumber
+    (OMathLog, TNumber)   -> returnA -< TNumber
+    (OMathCos, TNumber)   -> returnA -< TNumber
+    (OMathSin, TNumber)   -> returnA -< TNumber
+    (OMathAbs, TNumber)   -> returnA -< TNumber
+
+fresh :: ArrowState Location c => c () Location
+fresh = proc () -> do
+    Location s <- Control.Arrow.State.get -< ()
+    put -< Location $ s + 1
+    returnA -< Location $ s + 1
+
+getField_ :: ArrowChoice c => c (Type, String) Type'
+getField_ = proc (t, s) -> do
+    case t of
+        TObject fs -> case find (\(n, t) -> n == s) fs of
+            Just (n, t) -> returnA -< t
+            Nothing     -> returnA -< Data.Set.fromList [TUndefined]
+        _ -> returnA -< Data.Set.fromList [TObject [("0", Data.Set.singleton TString), ("length", Data.Set.singleton TNumber), ("$isArgs", Data.Set.singleton TBool)]]
+
+
+
+instance {-# OVERLAPS #-} AbstractValue Type' TypeArr where
+    -- values
+    numVal = proc _ -> returnA -< Data.Set.fromList [TNumber]
+    boolVal = proc _ -> returnA -< Data.Set.fromList [TBool]
+    stringVal = proc _ -> returnA -< Data.Set.fromList [TString]
+    undefVal = proc () -> returnA -< Data.Set.fromList [TUndefined]
+    nullVal = proc () -> returnA -< Data.Set.fromList [TNull]
+    lambdaVal = proc (ids, body) -> do
+        closure <- getEnv -< ()
+        returnA -< Data.Set.fromList [TLambda ids body closure]
+    objectVal = proc (fields) -> do
+        returnA -< Data.Set.fromList [TObject fields]
+    getField f1 = proc (subject, field) -> do
+        case field of
+            EString name -> returnA -< Prelude.foldr Data.Set.union (Data.Set.empty) (Prelude.map getField_ (zip (Data.Set.toList subject) (repeat name)))
+            _            -> returnA -< singleton TTop
+    updateField f1 = proc (_, _, _) -> returnA -< singleton TTop
+    deleteField f1 = proc (_, _) -> returnA -< singleton TTop
+    -- operator/delta function
+    evalOp = proc (op, vals) -> do
+        case length vals of
```

Comment:
Pattern match on `vals`. This is more efficient than calculating the length of the entire list.